libs/contrib/Data/String/Extra.idr

```diff
@@ -5,8 +5,12 @@ import Data.List1
 import Data.Nat
 import Data.String
 
-%default total
+%hide Data.String.lines
```

Comment:
What necessitated hiding these and redefining them? The comment below that says "should be removed once v.0.4.0 is released" makes the changes to this file particularly suspect :)

Reply:
Aye, I saw that too. I'd like to keep this as similar to the files in `src/Libraries` as possible, but this does seem like an awkward moment to copy that file. I'll try fix the duplication in `src/Libraries/Data/String/Extra.idr` to reduce the diff between it and its contrib counterpart. Though `libs/contrib/Data/String/Extra.idr` also has the functions `fastReplicate`, `justifyLeft`, `justifyRight`, which are absent from the `Libraries` version...

Reply:
This happens, when stuff is added to the standard libs that is also needed in the compiler. In these cases, the function in question is also added to an `Extra` module in `Libraries`, because it won't otherwise be available during bootstrapping. After a minor release, these temporary additions can be removed from the compiler libraries and the version from the standard libraries can be used.

Sorry for stating the obvious, if this was already known to you.

Reply:
Right, stuff in the contrib version but not in the compiler version of the Extra module are fine; that just means the functions have never been needed by the compiler. 

As Stefan noted, the compiler will duplicate functions prior to their availability in the previous release of the base libraries, but that consideration doesn't need to be made for the contrib package. In short, we should not need to hide and then re-export base library functions in contrib. 

The difference, IIUC, is in the order of compilation. The compiler is built using the prelude/base from the previous release and then _after_ that the new compiler is used to build the new version of prelude/base/contrib, etc.

Reply:
You are right, @mattpolzin . In this case, the problem with new stuff in the standard libs is actually not the bootstrap compiler (this will build the current version of the standard libs first) but the previous release, which will use the previous release version of the standard libs.

Reply:
https://github.com/idris-lang/Idris2/pull/1808/commits/e91a322d4184f53ed77d482673d1b8a69a373914 i pushed this commit as an attempt to remove the duplication from `Data.String.Extra`, it's currently failing on `Text.Literate` here: https://github.com/alexhumphreys/Idris2/blob/feat/sync-parsers/src/Libraries/Text/Literate.idr#L84 (and the same error in `contrib`). Changing `:::` to `::` causes a "not covering" error, and the function is kinda confusing so I'm not sure which cases or missing, or how the should be handled.

Thought I'd push this anyway, just to see if I'm on the right track with the de-duplication effort.

Reply:
Could this be related to #1585, where the semantics (and types) of `lines` was changed?

Reply:
Yep I think so. Now it probably has to handle an empty case which couldn't exist before.

Reply:
Or to be more precise:
```
Missing cases:
    with block in reduce _ [] _ _ _ _ _ _
```

Reply:
I suggest focusing on getting the contrib version working without redefining lines/unlines (using the new semantics called out by Stefan) and leave the task of getting the compiler version fixed fro a separate PR.

Because #1585 was merged _after_ version 0.4.0 was cut, removing the compiler redefinitions of lines/unlines will actually be a two-step process for all the same reasons discussed above:
1. actually update the redefined versions to align with the definitions in #1585 but leave them in plac
2. remove the redefinitions after the _next_ version of Idris is cut (0.5.0 or whatever it will be).

As noted above, we do not need to go through the same song and dance in contrib, we just need to use the semantics of the lines/unlines as defined in the current commit. Fixing remaining errors in contrib will likely still be the process @alexhumphreys is describing above of adding pattern clauses for the empty list case that is introduced by the new semantics.

Reply:
Aye, glad you said that as I was about to suggest the same. Even after the above error there's a bunch of changes to `src/Core` to be made. I just pushed a new commit to remove the duplication in contrib, but I duplicated the `lines` function in `Text.Literal` to preserve the old behaviour and not have to fiddle with that `reduce` function. Couldn't use a `where` clause for some reason, kept getting parse errors, so had to make a private namespace so the names didn't leak out.
Bit hacky, but I don't have much more time today so just wanted to get something reasonable up :D

Reply:
Thanks to gallais the tests should now be passing. I added the commit https://github.com/idris-lang/Idris2/pull/1808/commits/7fc48582160640dd873de3681f4da8f7e9f3c995 to explain that random `namespace` I added in `Text.Literal` and give it a better name. I was hoping to hide those functions in a `where` block but I don't think that's possible with `with`, and I've opened #1813 to track that.
Is there anything else stopping this from getting merged? If that `namespace` I added is too hacky, the `reduce` function could changed to use the new `lines`, but that'd probably need someone who understands it. I tried but wasn't able to grok it enough to be confident in changing it.